From 4f64c5eb6abb6bc75d3d49d2a0b409366068f0e7 Mon Sep 17 00:00:00 2001
From: Javier Martinez Canillas <javierm@redhat.com>
Date: Fri, 21 Jan 2022 18:10:25 +0100
Subject: [PATCH] nvidia-drm: remove existing generic drivers to take over the
 device

There are drivers that register framebuffer devices very early in the boot
process and make use of the existing framebuffer as setup by the firmware.

If one of those drivers has registered a fbdev, then the fallback fbdev of
the DRM driver won't be bound to the framebuffer console. To avoid that,
remove any existing generic driver and take over the graphics device.

By doing that, the fb mapped to the console is switched correctly from the
early fbdev (i.e: simpledrm) to the one registered by the real DRM drivers.

Signed-off-by: Javier Martinez Canillas <javierm@redhat.com>
---
 kernel/nvidia-drm/nvidia-drm-drv.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/kernel/nvidia-drm/nvidia-drm-drv.c b/kernel/nvidia-drm/nvidia-drm-drv.c
index 0d29d1d6bf4d..3f7725de6b91 100644
--- a/kernel/nvidia-drm/nvidia-drm-drv.c
+++ b/kernel/nvidia-drm/nvidia-drm-drv.c
@@ -60,6 +60,8 @@
 #include <drm/drm_ioctl.h>
 #endif
 
+#include <drm/drm_aperture.h>
+
 #include <linux/pci.h>
 
 /*
@@ -897,6 +899,8 @@ static void nv_drm_register_drm_device(const nv_gpu_info_t *gpu_info)
     struct drm_device *dev = NULL;
     struct device *device = gpu_info->os_device_ptr;
 
+    int ret = 0;
+
     DRM_DEBUG(
         "Registering device for NVIDIA GPU ID 0x08%x",
         gpu_info->gpu_id);
@@ -917,6 +921,16 @@ static void nv_drm_register_drm_device(const nv_gpu_info_t *gpu_info)
     mutex_init(&nv_dev->lock);
 #endif
 
+    /* Remove existing drivers that may own the framebuffer memory. */
+
+    ret = drm_aperture_remove_framebuffers(false, &nv_drm_driver);
+    if (ret) {
+	    NV_DRM_DEV_LOG_ERR(nv_dev,
+			       "Failed to remove existing framebuffers - %d.\n",
+			       ret);
+	    return;
+    }
+
     /* Allocate DRM device */
 
     dev = drm_dev_alloc(&nv_drm_driver, device);
-- 
2.34.1

